src/visitor.rs

```diff
@@ -47,64 +47,99 @@ impl Debug for NodeKind<'_, '_> {
     }
 }
 
+pub trait VisitResult {
+    fn stop_visit(&self) -> bool;
+}
+
+impl VisitResult for () {
+    fn stop_visit(&self) -> bool {
+        false
+    }
+}
+
+impl<X, E> VisitResult for Result<X, E> {
+    fn stop_visit(&self) -> bool {
+        self.is_err()
+    }
+}
+
+/// Visit the children of `self`, invoking `cb` on each.
+/// If `self` has no children, this method returns `None.
+/// Otherwise, it returns `Some(res)` where `res` is the return value of `cb`.
+#[allow(clippy::cyclomatic_complexity)]
 impl<'a, 't> NodeKind<'a, 't> {
-    pub fn for_each_child(&self, cb: &mut dyn FnMut(NodeKind<'a, 't>)) {
+    pub fn for_each_child<R>(&self, cb: &mut dyn FnMut(NodeKind<'a, 't>) -> R) -> Option<R>
+    where
+        R: VisitResult,
+    {
         use self::NodeKind::*;
         macro_rules! ccb {
-            ($astvar:expr) => {
-                cb(NodeKind::from($astvar))
-            };
+            ($astvar:expr) => {{
+                let res = cb(NodeKind::from($astvar));
+                if res.stop_visit() {
+                    return Some(res);
```

Comment:
?
```suggestion
                    return None;
```

Reply:
No, as described in the (misplaced) doc comment above, `for_each_child` returns None if no children were visited, otherwise the return value of the callback.